zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/LazyOpenInterpreter.java

```diff
@@ -32,6 +34,7 @@
     implements WrappedInterpreter {
   private Interpreter intp;
   volatile boolean opened = false;
+  private Lock lock = new ReentrantLock();
```

Comment:
Why use both lock and synchronized ?

Reply:
I think it can be changed like this?
```
  public void open() throws InterpreterException {
    if (lock.tryLock()) {
      try {
        if (opened == true) {
          return;
        } else {
          intp.open();
          opened = true;
        }
      } finally {
        lock.unlock();
      }
    }
  }
```
But now I don't understand why the 0.8 version of LazyInterpreter.open needs two synchronized?

Reply:
hmm, I think one synchronized is enough. 
It can be changed as 

```
  public void open() throws InterpreterException {
    if (opened == true) {
      return;
    }

    synchronized (intp) {
      if (opened == false) {
        intp.open();
        opened = true;
      }
    }
  }
```

Any reason using lock instead ?

Reply:
[create sparkContext thread] and [restart sparkInterpreter thread] compete for the same synchronized lock. If the SparkContext is not created, the spark interpreter will restart.
1. The phenomenon that the restart intereperter interface is stuck
2. On the basis of step 1, the user may kill -9 SparkSubmit process. At this time, the synchronized lock is released. [restart sparkInterpreter thread] acquires the lock and re-creates the sparkContext. **The phenomenon of re-appearing a spark app after killied a spark app**
   Because getSparkInterpreter() and LazyOpenInterpreter.cancel will be called to LazyOpenInterpreter.open(),I used lock.tryLock() instead of synchronized